File: javaNode/src/main/java/org/kevoree/library/command/AddInstance.java
package org.kevoree.library.command;

import org.kevoree.Instance;
import org.kevoree.KevoreeCoreException;
import org.kevoree.adaptation.AdaptationCommand;
import org.kevoree.adaptation.AdaptationType;
import org.kevoree.adaptation.KevoreeAdaptationException;
import org.kevoree.service.ModelService;
import org.kevoree.service.RuntimeService;
import org.kevoree.kcl.api.FlexyClassLoader;
import org.kevoree.library.InstanceRegistry;
import org.kevoree.library.wrapper.KInstanceWrapper;
import org.kevoree.library.wrapper.WrapperFactory;
import org.kevoree.log.Log;
import org.kevoree.modeling.api.KMFContainer;


/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 26/01/12
 * Time: 17:53
 */
public class AddInstance implements AdaptationCommand {

    private WrapperFactory wrapperFactory;
    private Instance instance;
    private InstanceRegistry registry;
    private RuntimeService runtimeService;
    private ModelService modelService;

    public AddInstance(WrapperFactory wrapperFactory, Instance instance, InstanceRegistry registry,
                       RuntimeService runtimeService, ModelService modelService) {
        this.wrapperFactory = wrapperFactory;
        this.instance = instance;
        this.registry = registry;
        this.runtimeService = runtimeService;
        this.modelService = modelService;
    }

    @Override
    public void execute() throws KevoreeAdaptationException {
        try {
            FlexyClassLoader fcl = runtimeService.installTypeDefinition(instance);
            Object instanceObject = runtimeService.createInstance(instance, fcl);
            KInstanceWrapper instanceWrapper = wrapperFactory.wrap(instance, instanceObject, runtimeService, modelService);
            instanceWrapper.setClassLoader(fcl);
//            runtimeService.injectDictionary(instance, instanceObject, true);

            registry.put(instance, instanceWrapper);
            Thread.currentThread().setContextClassLoader(null);
            Log.debug("Instance created {}", instance.path());
        } catch (KevoreeCoreException e) {
            throw new KevoreeAdaptationException("Unable to create instance " + instance.path(), e);
        }
    }

    @Override
    public void undo() throws KevoreeAdaptationException {
        new RemoveInstance(wrapperFactory, instance, registry, runtimeService, modelService).execute();
    }

    @Override
    public AdaptationType getType() {
        return AdaptationType.ADD_INSTANCE;
    }

    @Override
    public int hashCode() {
        return getType().hashCode() + instance.path().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof AdaptationCommand && obj.hashCode() == hashCode();
    }

    @Override
    public String toString() {
        return "AddInstance      " + instance.path();
    }

    @Override
    public KMFContainer getElement() {
        return instance;
    }
}

File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/defaultNodeTypes/command/AddInstance.kt
package org.kevoree.library.defaultNodeTypes.command

import org.kevoree.library.defaultNodeTypes.wrapper.KInstanceWrapper
import org.kevoree.library.defaultNodeTypes.ModelRegistry
import org.kevoree.Instance
import org.kevoree.api.BootstrapService
import org.kevoree.api.PrimitiveCommand
import org.kevoree.log.Log
import org.kevoree.library.defaultNodeTypes.wrapper.WrapperFactory
import org.kevoree.api.ModelService
import org.kevoree.ContainerNode

/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 26/01/12
 * Time: 17:53
 */

class AddInstance(val wrapperFactory: WrapperFactory, val c: Instance, val nodeName: String, val registry: ModelRegistry, val bs: BootstrapService, val modelService: ModelService) : PrimitiveCommand, Runnable {

    var nodeTypeName: String? = null
    var tg: ThreadGroup? = null

    var resultSub = false

    override fun execute(): Boolean {
        var subThread: Thread? = null
        try {
            tg = ThreadGroup("kev/" + c.path()!!)
            subThread = Thread(tg, this)
            subThread!!.start()
            subThread!!.join()
            return resultSub
        } catch(e: Throwable) {
            if (subThread != null) {
                try {
                    subThread!!.stop() //kill sub thread
                } catch(t: Throwable){
                    //ignore killing thread
                }
            }
            val message = "Could not add the instance " + c.name!! + ":" + c.typeDefinition!!.name!!
            Log.error(message, e)
            return false
        }
    }

    override fun undo() {
        RemoveInstance(wrapperFactory, c, nodeName, registry, bs, modelService).execute()
    }

    public override fun run() {
        try {
            var newBeanKInstanceWrapper: KInstanceWrapper? = null
            if (c is ContainerNode) {
                newBeanKInstanceWrapper: KInstanceWrapper? = wrapperFactory.wrap(c, this/* nodeInstance is useless because launched as external process */, tg!!, bs, modelService)
                registry.register(c, newBeanKInstanceWrapper!!)
            } else {
                val newBeanInstance = bs.createInstance(c)
                newBeanKInstanceWrapper: KInstanceWrapper? = wrapperFactory.wrap(c, newBeanInstance!!, tg!!, bs, modelService)
                registry.register(c, newBeanKInstanceWrapper!!)
                bs.injectDictionary(c, newBeanInstance, true)
            }
            newBeanKInstanceWrapper?.create()
            resultSub = true
        } catch(e: Throwable){
            Log.error("Error while adding instance {}", e, c.name)
            resultSub = false
        }
    }


    public override fun toString(): String {
        return "AddInstance " + c.name
    }
}